### PR TITLE
[ChatStateLayer] Share mark and mark unread message

### PR DESCRIPTION
**Pull request by laevandus**
### 🔗 Issue Links

Related: [#728](https://github.com/GetStream/ios-issues-tracking/issues/728)

*Merges to feature/chat-state-layer*

### 🎯 Goal

Share the logic used for marking channel read.

### ☑️ Contributor Checklist

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] This change follows zero ⚠️ policy (required)
- [ ] This change should be manually QAed
- [ ] Changelog is updated with client-facing changes
- [x] New code is covered by unit tests
- [ ] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (docusaurus, tutorial, CMS)

### CODE REVIEW

File: Sources/StreamChat/Workers/ReadStateSender.swift [hunk 1]
```diff
@@ -4,74 +4,96 @@
...
-            cid: ChannelId,
-            channelUpdater: ChannelUpdater,
-            authenticationRepository: AuthenticationRepository,
-            messageRepository: MessageRepository
-        ) {
-            self.authenticationRepository = authenticationRepository
-            self.cid = cid
-            self.channelUpdater = channelUpdater
-            self.messageRepository = messageRepository
+final class ReadStateSender {
```
Comment (laevandus) [inline Sources/StreamChat/Workers/ReadStateSender.swift:7]: Moved controller's logic to this type which I created for the state layer. Now both use the same logic.

File: Sources/StreamChat/Workers/ReadStateSender.swift [hunk 2]
```diff
@@ -4,74 +4,96 @@
...
-                    try await channelUpdater.markRead(cid: cid, userId: currentUserId)
+        }
+    }
+    
+    func markUnread(
+        from messageId: MessageId,
+        in channel: ChatChannel,
+        completion: @escaping (Result<ChatChannel, Error>) -> Void
+    ) {
+        guard !markingRead,
```
Comment (martinmitrevski) [inline Sources/StreamChat/Workers/ReadStateSender.swift:57]: I guess it's ok we use the same bool for marking as unread, do you see any issues with it? (seems like it was like that before too)

File: Tests/StreamChatTests/Controllers/ChannelController/ChannelController_Tests.swift [hunk 3]
```diff
@@ -4182,16 +4182,16 @@ final class ChannelController_Tests: XCTestCase {
         }
         client.setToken(token: .unique(userId: .unique))
 
+        let updater = try XCTUnwrap(env.channelUpdater)
```
Comment (martinmitrevski) [inline Tests/StreamChatTests/Controllers/ChannelController/ChannelController_Tests.swift:4185]: why was this needed?

**Review by martinmitrevski — APPROVED** [review]

File: Sources/StreamChat/Workers/ReadStateSender.swift (hunk 1 above)
Comment (nuno-vieira) [inline Sources/StreamChat/Workers/ReadStateSender.swift:7]: I would rename it to something like this:
```suggestion
final class ReadStateHandler {
```

The "Sender" postfix is used as a background worker, like the "MessageSender", so it is used fro components which are observing data changes and then queue and send requests async. 

This component the goal is just to reuse the State logic, kinda like we have the PaginationStateHandler. So I guess it makes sense to have consistancy and also have the "Handler" postfix

File: Sources/StreamChat/Workers/ReadStateSender.swift (hunk 2 above)
Comment (nuno-vieira) [reply Sources/StreamChat/Workers/ReadStateSender.swift:57]: I think it is important we use the same bool. We don't want to mark **unread**, while we are marking **read**, otherwise we will have race conditions. The problem is that it seems that we maybe lose some requests here. If it happens a user marks unread/read while one of the requests is being executed we don't do anything

Comment (nuno-vieira) [reply Sources/StreamChat/Workers/ReadStateSender.swift:57]: But I would say that is not a big deal at the moment

Comment (laevandus) [reply Sources/StreamChat/Workers/ReadStateSender.swift:57]: Currently I just replicated what controller was doing and did not want to change any logic there. But that said, it does not seem like 100% reliable implementation when many unread and read calls happen.

File: Sources/StreamChat/Workers/ReadStateSender.swift (hunk 1 above)
Comment (laevandus) [reply Sources/StreamChat/Workers/ReadStateSender.swift:7]: Sure, let's do that. I'll do another commit and fix all the call sites.

File: Tests/StreamChatTests/Controllers/ChannelController/ChannelController_Tests.swift (hunk 3 above)
Comment (laevandus) [reply Tests/StreamChatTests/Controllers/ChannelController/ChannelController_Tests.swift:4185]: This was because I needed to move the logic how we fetch the last read message. Before it uses observer, now it fetches from the DB which caused a delay and I needed to change tests to take this into account.
